File: py/combinations/permutations.py
def permutations(str):
    def swap(arr, i, j):
        arr[i], arr[j] = arr[j], arr[i]

    def _perm(l, start, end):
        if start == end:
            print(l)
        else:
            for i in range(start, end):
                swap(l, start, i)
                _perm(l, start+1, end)
                swap(l, start, i)

    l = list(str)
    _perm(l, 0, len(l))

permutations("abc")

File: java/src/snippets/Permutations.java
package snippets;

public class Permutations {

    public static void swap(char[] a, int i, int j){
        char temp = a[i];
        a[i] = a[j];
        a[j] = temp;
    }

    public static void permutations(char [] a, int pos){
        if (pos == a.length)
            System.out.println(new String(a));
        else {
            for (int i = pos; i < a.length; i++) {
                swap(a, pos, i);
                permutations(a, pos + 1);
                swap(a, pos, i);
            }
        }
    }

    public static void main(String[] args){
        char [] c = {'a', 'b', 'c', 'd'};
        permutations(c, 0);
    }
}